pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,231 @@
+			level.Info(logger).Log("msg", "circuit breaker is closed", "previous", event.OldState, "current", event.NewState)
+		}).
+		OnOpen(func(event circuitbreaker.StateChangedEvent) {
+			circuitBreakerTransitionsCounterFn(cb.metrics, circuitbreaker.OpenState).Inc()
+			level.Warn(logger).Log("msg", "circuit breaker is open", "previous", event.OldState, "current", event.NewState)
+		}).
+		OnHalfOpen(func(event circuitbreaker.StateChangedEvent) {
+			circuitBreakerTransitionsCounterFn(cb.metrics, circuitbreaker.HalfOpenState).Inc()
+			level.Info(logger).Log("msg", "circuit breaker is half-open", "previous", event.OldState, "current", event.NewState)
+		})
+
+	if cfg.testModeEnabled {
+		// In case of testing purposes, we initialize the circuit breaker with count based failure thresholding,
+		// since it is more deterministic, and therefore it is easier to predict the outcome.
+		cbBuilder = cbBuilder.WithFailureThreshold(cfg.FailureThresholdPercentage)
+	} else {
+		// In case of production code, we prefer time based failure thresholding.
+		cbBuilder = cbBuilder.WithFailureRateThreshold(cfg.FailureThresholdPercentage, cfg.FailureExecutionThreshold, cfg.ThresholdingPeriod)
+	}
+
+	cb.cb = cbBuilder.Build()
+	cb.metrics = newCircuitBreakerMetrics(registerer, cb.cb.State)
+	return &cb
+}
+
+func isFailure(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// We only consider timeouts or ingester hitting a per-instance limit
+	// to be errors worthy of tripping the circuit breaker since these
+	// are specific to a particular ingester, not a user or request.
+
+	if errors.Is(err, context.DeadlineExceeded) {
+		return true
+	}
+
+	statusCode := grpcutil.ErrorToStatusCode(err)
+	if statusCode == codes.DeadlineExceeded {
+		return true
+	}
+
+	var ingesterErr ingesterError
+	if errors.As(err, &ingesterErr) {
+		return ingesterErr.errorCause() == mimirpb.INSTANCE_LIMIT
+	}
+
+	return false
+}
+
+func (cb *circuitBreaker) isActive() bool {
+	if cb == nil {
+		return false
+	}
+	return cb.active.Load()
+}
+
+func (cb *circuitBreaker) setActive() {
+	if cb == nil {
+		return
+	}
+	cb.active.Store(true)
+}
+
+// tryAcquirePermit tries to acquire a permit to use the circuit breaker and returns whether a permit was acquired.
+// If it was possible to acquire a permit, success flag true and no error are returned. The acquired permit must be
+// returned by a call to finishPushRequest.
+// If it was not possible to acquire a permit, success flag false is returned. In this case no call to finishPushRequest
+// is needed. If the permit was not acquired because of an error, that causing error is returned as well.
+func (cb *circuitBreaker) tryAcquirePermit() (bool, error) {
+	if !cb.isActive() {
+		return false, nil
+	}
+	if !cb.cb.TryAcquirePermit() {
+		cb.metrics.circuitBreakerResults.WithLabelValues(resultOpen).Inc()
+		return false, newCircuitBreakerOpenError(cb.cb.RemainingDelay())
+	}
+	return true, nil
+}
```

Comment:
I would expect that if `tryAcquirePermit` returns `false, nil`, then I don't have permit to proceed. However, it's not like that.

I would try to refactor the method to either avoid having to read the comment to understand the return values (what's the first return value? prod code calls it "acquired" while tests call it "status") or at least to force the user to think about it.

The comment above says:
> If it was not possible to acquire a permit, success flag false is returned. In this case no call to finishPushRequest is needed.

Which puts logic in the caller of this method who shouldn't care whether it's active or not, and I think we can do that better by just returning a callback:

```suggestion
func (cb *circuitBreaker) tryAcquirePermit() (finish func(duration time.Duration, pushErr error), err error)
	if !cb.isActive() {
		return func(_ time.Duration, _ error,) {}, nil
	}
	if !cb.cb.TryAcquirePermit() {
		cb.metrics.circuitBreakerResults.WithLabelValues(resultOpen).Inc()
		return nil, newCircuitBreakerOpenError(cb.cb.RemainingDelay())
	}
	return cb.finishPushRequest, nil
}
```

Then the caller doesn't have more logic than just:

```go
finish, err := cb.tryAquirePermit()
if err != nil {
    return ..., err
}
// ...
finish(time.Since(t0), err)
```

Reply:
While I understand your rationale (client just calls returned function), it looks like slightly over-engineered solution to a trivial problem, where a simple boolean works just fine.

Reply:
I would also keep the current implementation as it is.

Reply:
> it looks like slightly over-engineered solution

Closures are first class citizens just like bools :)

> to a trivial problem

The difference is that it removes the problem of having to decide things from the caller.